node-repository/src/main/java/com/yahoo/vespa/hosted/provision/Node.java

```diff
@@ -45,37 +46,29 @@
+    public static Node createDockerNode(Set<String> ipAddresses, Set<String> ipAddressPool, String hostname, String parentHostname, Flavor flavor, NodeType type) {
+        return new Builder("fake-" + hostname, ipAddresses, hostname, flavor, type)
+                .withIpAddressPool(ipAddressPool)
+                .withParentHostname(parentHostname)
+                .withState(State.reserved)
+                .build();
     }
 
-    /** Creates a node in the initial state (reserved for docker containers, provisioned otherwise) */
+    /** Creates a node in the initial state (provisioned) */
     public static Node create(String openStackId, Set<String> ipAddresses, Set<String> ipAddressPool, String hostname, Optional<String> parentHostname, Flavor flavor, NodeType type) {
-        return new Node(openStackId, ipAddresses, ipAddressPool, hostname, parentHostname, flavor, Status.initial(), State.provisioned,
-                        Optional.empty(), History.empty(), type);
-    }
-
-    /** Do not use. Construct nodes by calling {@link NodeRepository#createNode} */
-    private Node(String id, Set<String> ipAddresses, Set<String> ipAddressPool, String hostname, Optional<String> parentHostname,
-                 Flavor flavor, Status status, State state, Allocation allocation, History history, NodeType type) {
-        this(id, ipAddresses, ipAddressPool, hostname, parentHostname, flavor, status, state, Optional.of(allocation), history, type);
+        return new Builder(openStackId, ipAddresses, hostname, flavor, type)
+                .withIpAddressPool(ipAddressPool)
+                .withParentHostname(parentHostname)
+                .build();
     }
 
+    /**
+     * Creates a node with all fields specified, necessary for serialization code.
+     *
+     * Others should use the {@code create} methods to create nodes, or the Builder to modify nodes.
+     */
     public Node(String id, Set<String> ipAddresses, Set<String> ipAddressPool, String hostname, Optional<String> parentHostname,
-                Flavor flavor, Status status, State state, Optional<Allocation> allocation,
-                History history, NodeType type) {
-        Objects.requireNonNull(id, "A node must have an ID");
-        requireNonEmpty(ipAddresses, "A node must have at least one valid IP address");
-        requireNonEmptyString(hostname, "A node must have a hostname");
-        requireNonEmptyString(parentHostname, "A parent host name must be a proper value");
-        Objects.requireNonNull(flavor, "A node must have a flavor");
-        Objects.requireNonNull(status, "A node must have a status");
-        Objects.requireNonNull(state, "A null node state is not permitted");
-        Objects.requireNonNull(allocation, "A null node allocation is not permitted");
-        Objects.requireNonNull(history, "A null node history is not permitted");
-        Objects.requireNonNull(type, "A null node type is not permitted");
```

Comment:
I think the contracts should be kept here as the `Node` class should have the final say in what's allowed in its fields or not.

Reply:
Agree. It also provides a simple to find and unambiguous answer to the question "what values can this field have?".

Reply:
Done